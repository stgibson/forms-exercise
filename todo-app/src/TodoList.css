.TodoList {
  margin: 20px;
}