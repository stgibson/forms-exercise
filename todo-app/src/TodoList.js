import React, { useState } from "react";
import Todo from "./Todo";
import NewTodoForm from "./NewTodoForm";
import { v4 as uuid } from "uuid";
import "./TodoList.css";

const TodoList = () => {
  const [todos, setTodos] = useState([]);

  const addTodo = task => setTodos(() => [ ...todos, { id: uuid(), task }]);

  const removeTodo = event => {
    const { id } = event.target.parentElement;
    setTodos(() => todos.filter(todo => todo.id !== id))
  };

  return (
    <div className="TodoList">
      <NewTodoForm addTodo={ addTodo } />
      {
        todos.map(todo => (
          <Todo
            id={ todo.id }
            key={ todo.id }
            task={ todo.task }
            removeTodo={ event => removeTodo(event) }
          />
        ))
      }
    </div>
  );
};

export default TodoList;