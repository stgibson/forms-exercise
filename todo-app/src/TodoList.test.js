import React from "react";
import TodoList from "./TodoList";
import { render, fireEvent } from "@testing-library/react";

it("renders without crashing", () => {
  render(<TodoList />);
});

it("matches snapshot", () => {
  const { asFragment } = render(<TodoList />);
  expect(asFragment()).toMatchSnapshot();
});

it("updates form when user types in task field", () => {
  const { getByLabelText, queryByDisplayValue } = render(<TodoList />);

  const input = getByLabelText("Task:");

  fireEvent.change(input, { target: { name: "task", value: "Test Task" } });

  expect(queryByDisplayValue("Test Task")).toBeInTheDocument();
});

it("can add a new todo", () => {
  const { getByLabelText, queryByText } = render(<TodoList />);

  const input = getByLabelText("Task:");
  const submitButton = queryByText("Add Todo");

  fireEvent.change(input, { target: { name: "task", value: "Test Task" } });
  fireEvent.click(submitButton);

  expect(queryByText("Test Task")).toBeInTheDocument();
});

it("can delete a todo", () => {
  const { getByLabelText, queryByText } = render(<TodoList />);

  const input = getByLabelText("Task:");
  const submitButton = queryByText("Add Todo");

  // first add a todo
  fireEvent.change(input, { target: { name: "task", value: "Test Task" } });
  fireEvent.click(submitButton);

  // since only 1 todo, find button with X and click to delele
  const deleteButton = queryByText("X");
  fireEvent.click(deleteButton);

  expect(queryByText("Test Task")).not.toBeInTheDocument();
});