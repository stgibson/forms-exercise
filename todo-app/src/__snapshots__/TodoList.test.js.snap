// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="TodoList"
  >
    <form
      class="NewTodoForm"
    >
      <label
        for="task"
      >
        Task:
      </label>
      <input
        class="NewTodoForm-input"
        id="task"
        name="task"
        type="text"
        value=""
      />
      <button
        type="submit"
      >
        Add Todo
      </button>
    </form>
  </div>
</DocumentFragment>
`;
